### ISSUE TITLE
[C+ Checklist Needs Completion] [$500] Web - The app is freezing after switching tabs, and the performance of the tab navigator seems to be quite poor

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#31190]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.98.0
**Reproducible in staging?:** y
**Reproducible in production?:** n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:**  @situchan 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1699629011502749?thread_ts=1699456424.016869&cid=C01GTK53T8Q

## Action Performed:
1. Go to staging.new.expensify.com
2. Tab on Fab menu
3. Select Request Money
4. Navigate through the tabs

## Expected Result:
 The app is not freezing after switching tabs, and the performance of the tab navigator should be good
 
## Actual Result:
 The app is freezing after switching tabs, and the performance of the tab navigator seems to be quite poor

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/1e304330-c689-4e09-8fbd-8b627a8e817b

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c6d30dd934f349b1</li>
        <li>Upwork Job ID: 1723032823278329856</li>
        <li>Last Price Increase: 2023-11-10</li>
    </ul>
</details>

**Comment by kbecciv** [comment]
Removing Deploy blocker per the comment  https://expensify.slack.com/archives/C01GTK53T8Q/p1699637937484459?thread_ts=1699456424.016869&cid=C01GTK53T8Q

**Comment by greg-schroeder** [comment]
Awaiting proposals

**Comment by davidnormo** [comment]
Contributor details
Your Expensify account email: david.normo@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0108d5ebd2889d3ebf

**Comment by davidnormo** [comment]
@mananjadhav @greg-schroeder @stitesExpensify what do you guys think?

**Comment by mananjadhav** [comment]
I'll review this later today.

**Comment by mananjadhav** [comment]
Thanks for the proposal @davidnormo. I can se you've mentioned about `TabNavigator.web.js`. I can't identify how we derived to this conclusion.

> The animations are being run by @react-navigation/material-top-tabs which uses react-native's animated library. I can't find any extra state that is changing that is causing additional work. It just appears that the application is too "heavy" for this animation on web.

Also the root cause seems to be unclear. Can we be specific about what's causing the delay? and then also elaborate on the solution?


@stitesExpensify @mountiny do you think this is a good candidate for the #vip-newdot-performance ?

**Comment by situchan** [comment]
This is recent regression. First of all, we should find offending PR.

**Comment by mountiny** [comment]
Yep I think this would be great for the Performance project

I think this started with the RNW upgrade

**Comment by greg-schroeder** [comment]
Not sure who is working on this - is this going to be worked on internally or no?

**Comment by stitesExpensify** [comment]
@greg-schroeder it will be worked on externally as part of that VIP project.

**Comment by greg-schroeder** [comment]
Waiting for proposals then

**Comment by stitesExpensify** [comment]
@mountiny do we need to do anything to get this picked up besides having it be a part of the GH project?

**Comment by mountiny** [comment]
@stitesExpensify We can call it out in the performance room. I will do it

**Comment by kacper-mikolajczak** [comment]
Hi, I am Kacper from Callstack and I would like to investigate that issue!

**Comment by greg-schroeder** [comment]
Okay assigned to @kacper-mikolajczak, not overdue Melvy

**Comment by greg-schroeder** [comment]
Draft PR is up

**Comment by mananjadhav** [comment]
I am not sure why overdue label was added when @greg-schroeder posted a response yesterday. Anyway the draft PR is up, and will start reviewing once it is ready.

**Comment by greg-schroeder** [comment]
Yeah, not sure why Melvin is getting so touchy

**Comment by kacper-mikolajczak** [comment]
Hi folks! As an explainer to the [PR](https://github.com/Expensify/App/pull/32577), I wanted to post a quick summary in order to be on the same page with introduced changes.

### **Problem**
Request money screen transitions lag heavily on web.

### **Investigations**
For web, `react-navigation` and `react-native-web` use `Animated` which does not support web's native animation APIs. In order for animations to work, the library updates react values via `requestAnimationFrame` that schedules the update on each frame. Unfortunately, this means that many commits are triggered on every frame, which beyond a certain point of screen complexity are not possible to be processed in a meaningful time frame.

Currently, there are two animations taking place - one is regular screen's content horizontal translation and the second is responsible for animating `TabBar` properties. In this investigation I've tried to find out which parts of the `MoneyRequestSelectorPage` (a component behind Request money) impact the performance the most.

All of the measurements were made when navigating between `Manual` and `Scan` screens without CPU throttling at 165Hz refresh rate, but what matters the most is the general pattern and relative gains, not necessarily absolute values.

#### _Main branch (default)_
The main branch is used as a representative sample of the current state. Here are the profiling results of navigating:

- number of commits: **~4k**
- frame time: **50-70ms**

An `onUpdate` from `Animated` is being called many times in a single `rAF` invocation, which is triggering recursive calls of `updateFiberRecursively`.

![Screenshot 2023-12-11 at 12 31 11](https://github.com/Expensify/App/assets/62747088/093d4b61-50ea-49a8-9b82-fd3b3b83f7c2)

#### _Animations disabled (baseline)_
As we can see from the default profiling trace, there is a lot going on in each frame. In order to see if animations are the root cause, they were disabled in this run, which will serve as a baseline for further improvements:

- number of commits: **36**
- profiling trace time: **~150ms**

![Screenshot 2023-12-11 at 14 24 33](https://github.com/Expensify/App/assets/62747088/1638300f-9900-4b2f-b27e-8d7d040e7619)
![Screenshot 2023-12-11 at 14 28 42](https://github.com/Expensify/App/assets/62747088/10d6c74a-08d3-42a4-9d36-0be948d652c5)

#### _No content_
In order to rule out the impact of the tabs content, I rendered them without any children (`() => null`). The results tell that it is not the main issue.

![Screenshot 2023-12-11 at 12 43 02](https://github.com/Expensify/App/assets/62747088/aa97114d-550d-4627-8889-e3716fa54260)

#### _Default TabBar_
`ReactNavigation` comes with a default `TabBar` component, that was used in place the custom one to see the difference. It looks like the our custom bar `TabSelector` is quite impactful. It is probably due to the heavy animations that are taking place whenever we switch the tabs.

![image](https://github.com/Expensify/App/assets/62747088/d38ea81e-2ab5-4a15-968a-90f4b4a05f8f)

### **Solution**

The problem seems to be going deeper than the app itself, although there is a lot to cover here as well. With that in mind, I've reached out to the React Navigation folks to understand what we can do to fix the root cause.

In the meantime, I would suggest temporarily **disabling** `Tabs` animations on **web-based platforms**. It would not cut out any functionality while significantly helping with UX.

I am attaching the recordings with comparison of how the app behaves with animations disabled.
The fix in its current form is available in the [PR](https://github.com/Expensify/App/pull/32577) with addition of a few tweaks to how the animations are being handled.

https://github.com/Expensify/App/assets/62747088/2101102d-5829-45bf-9477-b3803f4de11c


https://github.com/Expensify/App/assets/62747088/58611105-72e8-4c13-b75e-96b16ccf4389

As always, I am open to your ideas and feedback - let me know what you think. Thanks a lot! 👍

**Comment by abzokhattab** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Transitions between Tabs become laggy after a few clicks

### What is the root cause of that problem?

In the [new chat selector page](https://github.com/Expensify/App/blob/e87d35d8af177df33c4fdfbf221ead732efec310/src/pages/NewChatSelectorPage.js#L39-L57) we use the `TabSelector` component  as a prop inside the `OynxTabNavigator` component 

This `TabSelector` gets rerendered everytime we click on a tab so after some clicks, the transitions become very slow.

This issue affects all components that use the `TabSelector`

1. `NewChatSelectorPage`
2. `MoneyRequestSelectorPage`
3. `IOURequestStartPage`

### What changes do you think we should make in order to solve the problem?

we need to use `React.memo` on the `TabSelectorItem`:
 
to do so we need to add this line:

```js
    const MemoizedTabSelectorItem = React.memo(TabSelectorItem);
```

and change this 
https://github.com/Expensify/App/blob/fa87f8bb98aa960f71584e188cbaee5c091bb0f0/src/components/TabSelector/TabSelector.js#L131
to 
```js
<MemoizedTabSelectorItem
```

**Comment by greg-schroeder** [comment]
Linked PR is in final review stages, @mananjadhav you're up!

**Comment by mananjadhav** [comment]
Yes currently reviewing the code.

**Comment by abzokhattab** [comment]
Hello team just wondering if you are still accepting proposals here and i just noticed that my [proposal](https://github.com/Expensify/App/issues/31190#issuecomment-1856591632) was hidden to off topic. 
can you please elaborate why is off topic as far as I understand the seamless navigation is the purpose here and I attached a video for it. Just wanted to know 

Thanks

**Comment by greg-schroeder** [comment]
This issue already has an assigned contributor and PR in review, so we’re
not accepting any new proposals. That’s why I marked the proposal as off
topic.


Website <https://we.are.expensify.com/> | Newsletter
<https://we.are.expensify.com/media>

**Comment by abzokhattab** [comment]
Thanks @greg-schroeder  for the reply, I see that in the PR we are mentioning that we are disabling the animation till the root cause is addressed,
 > This PR aims to refactor animations and temporarily disable them on web-based platforms until root cause is addressed in React Navigation

 I think the solution that I raised mentions the root cause of the laggy animations and solves it without disabling the animation, that's why it might be worthy to assess it (maybe in another issue since the current issue is already assigned to a contributer).

**Comment by greg-schroeder** [comment]
@abzokhattab Thanks for your feedback! I think we're happy with the direction we're going for now but certainly keep your eyes peeled for related issues labeled `External` and `Help Wanted` to contribute to in the future.

**Comment by shubham1206agra** [comment]
Pu this on hold for https://github.com/Expensify/App/pull/32764

**Comment by mananjadhav** [comment]
@kacper-mikolajczak Are we making changes in the PR? Let me know when this is ready for review again.

**Comment by kacper-mikolajczak** [comment]
> @kacper-mikolajczak Are we making changes in the PR? Let me know when this is ready for review again.

Hi @mananjadhav, PR is under review here https://github.com/Expensify/App/pull/33841

Pretty me know if need any help

**Comment by greg-schroeder** [comment]
This is not being paid today given the deploy blocker

**Comment by greg-schroeder** [comment]
Okay so there was another deploy blocker related to this? Seems https://github.com/Expensify/App/issues/34377 is still open...

**Comment by mananjadhav** [comment]
@greg-schroeder I wouldn't call it a deploy blocker because we intentionally disabled animations. 

wdyt @mountiny @stitesExpensify ?

Also this is ready for payout.

**Comment by shubham1206agra** [comment]
@greg-schroeder I think I am up for partial compensation.

@mountiny can verify this.

**Comment by mananjadhav** [comment]
> @greg-schroeder I think I am up for partial compensation.
> 
> @mountiny can verify this.

Agreed with @shubham1206agra here.


I don't think we have any offending PR as this was a performance issue, and we disable the animation as a solution. Neither do I think we need a regression test for this one.

**Comment by mountiny** [comment]
@mananjadhav @shubham1206agra I can see you both have reviewed the PR. I agree there was no regression, but I feel like you should divide the reward of $500 in this case, its not clear to me now why both c+ were needed

**Comment by mananjadhav** [comment]
@mountiny More info [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1703852548439579?thread_ts=1703843823.921549&cid=C01GTK53T8Q) on why @shubham1206agra got involved and where we committed for the partial payout.

I think $250 each is fine for me.

**Comment by mountiny** [comment]
Thank you for the link, I agree $250 each sounds fair!

**Comment by mananjadhav** [comment]
@greg-schroeder Can you add a payment summary comment for us to raise a request?

**Comment by JmillsExpensify** [comment]
$250 payment approved based on [this comment](https://github.com/Expensify/App/issues/31190#issuecomment-1912524954). That is sufficient.

**Comment by greg-schroeder** [comment]
Sorry, yeah okay so the final breakdown was:

@mananjadhav $250
@shubham1206agra $250

**Comment by shubham1206agra** [comment]
@greg-schroeder I didn't received any payment for the same.

**Comment by greg-schroeder** [comment]
Sorry, this was confusing @shubham1206agra - paying you here: https://github.com/Expensify/App/issues/35793

**Comment by shubham1206agra** [comment]
@greg-schroeder Accepted offer.
